### DISCUSSION TITLE
VAT Number validation (EU)

### DISCUSSION THREAD

**Discussion by robertsimon** [nav-gov-hu/eVAT#28] [Kérdések és válaszok - Gépi interfész]
Sziasztok!

A NAV-nak van valamiféle API-ja, ami *NEM* a VIES-t használja EU-s VAT szám ellenőrzéshez? 
Jó sokat kéne validálnunk (napi 5k-200k) és a VIES egy retek, használhatatlan realtime/semi-realtime. 
Előre is köszönöm az építő jellegű segítséget. 

R.

**Comment by NTCA-developer** [comment]
Szia @robertsimon 

Első körben hadd kérdezzek vissza!
- Ha a bemenet a nemzetiség és az adószám, illetve a kimenet egy logikai boolean (jó/nem jó) az elég-e nektek üzletileg? Mire, milyen folyamaton belül használjátok ezeket a validációkat?
- Tudtok-e olyan használati esetet mondani, ami indokolná azt hogy egy ilyen feltételezett végpont az eÁFA API alá kerüljön?

**Reply by robertsimon** [reply]
Üdv, elnézést, hogy csak most írok, kb 5 napig szenvedtem a VIES-el, amire levalidáltam DE, FR, BE, IT, PT, PL és HU adószámokat. 
Bemenetnként bőven elég egy ilyen: 
`public class VatNumberValidationResponse
{
	public string VatNumber { get; set; }
	public string CountryCode { get; set; }
}`

A response pedig lehet vmi ilyesmi: 
`public class VatNumberValidationResponse
{
	public string VatNumber { get; set; }
	public string CountryCode { get; set; }
	public bool IsValidFormat { get; set; }
	public bool? IsValid { get; set; }
	public int? ErrorCode { get; set; }
}`

Nagyon nagyon előnyös lenne ha lehetne mass-validation-t is csinálni, tehát beküldök egy rakás adószámot (mondjuk max 10k-t) és visszakapom a validált adószámokat response-ba. 
A legnagyobb csillantás az lenne, ha lenne módotok csekkolni az adószámokat direktbe a többi ország TA-ján (Tax Authority) keresztül. 

Második kérdésedre az alábbi kollégák szerintem mondtak pár választ.
Nagyon szépen köszönöm az idődet.

R.

**Reply by robertsimon** [reply]
Még adós maradtam egy válasszal: 
"Mire, milyen folyamaton belül használjátok ezeket a validációkat?"
Nemzetközi ügyfelek tranzakcióit normalizáljuk, validáljuk és készítünk belőle adóbevallásokat, amiket utána be is adunk az EU adóhatóságaihoz.

**Reply by NTCA-developer** [reply]
Semmi gond, köszönöm a válaszokat!

Na pont ezért kérdeztem vissza, mert közben itt többünknek elkezdett bilibe lógni a keze, hogy úgy mondjam. :) Kérek szépen pár napot hogy egyeztessek az illetékesekkel, visszajelzek.

**Comment by vectorkovacspeter** [comment]
Csatlakozom a kérdezőhöz, jelenleg nincs igazán jó m2m mód erre, [vatapi.com](https://vatapi.com/) és [vatlayer.com](https://vatlayer.com/) fizetős szolgáltatások váltak be nálunk, de jobb lenne ha támogatná ezt a rendszer.

Az igen/nem válasz szerintem a minimálisan elégséges szint, de klassz lenne cégnév, cím is.

Használati esetként én azon adózók támogatását említeném, akik nemzetközi ügyleteket is bonyolítanak, és a jó adózói magatartás illetve a visszaélések és adminisztrációs hibák elkerülése érdekében validálnák ezen adatokat is

**Comment by janosnad** [comment]
Ez talán az onlineszámla /queryTaxpayer mellé férne be a legjobban, hiszen a számlázás előtt kellene ezt is ellenőrizni. 
Felhasználói oldalon nagy igény volna rá, mert sokan a számlát sem akarják addig kiállítani, amíg nem kapnak megerősítést az adószám hitelességéről.
Persze csak akkor van értelme, ha a https://nav.gov.hu/ugyfeliranytu/eu_adoszam alatt olvasható alapfeltételt tudja biztosítani:
"Amikor megszületik az eredmény, Ön egy egyedi hivatkozási számot kap, mely alapján a NAV az Ön által kezdeményezett lekérdezést – ha az ellenőrzött közösségi adószám érvényes volt – azonosítani tudja. Fontos, hogy ezt az egyedi hivatkozási számot megőrizze, mert így tudja igazolni a lekérdezés tényét és eredményét."
Sajnos az itt javasolt webszolgáltatás megbízhatatlan. Tapasztalat szerint az ugyanazon IP-rő beérkező sok kérés esetén dobja el magát véletlenszerűen.

**Reply by vectorkovacspeter** [reply]
Onlineszámálba sem lenne rossz :)
A vies ami használhatatlan, de a vatapi -t melegen ajánlom, és ad hivatkozási számot is, sőt azt is megmondja, hogy milyen illetékes hatósághoz tartozik a keresett adószám. Csak sajnos extra pénzbe kerül a használata...

**Reply by robertsimon** [reply]
Megnéztem a VatApi-t, UK-hez a HRMERCI-t használják, EU-hoz meg a VIES-t. Ugyanezt a svc-t csináltuk meg mi is. 
https://docs.vatapi.com/#rate-limit

"Validate UK & EU member state VAT numbers, this resource queries HMRC for UK VAT numbers and the VIES database ran by the European Commission for EU member state VAT numbers."

**Comment by robertsimon** [comment]
Nagyon szépen köszönöm válaszaitokat! 

Mi csináltunk egy olyan service-t,  ami REST API-n keresztül befogad x darab adószámot, amit megpróbál utána validálni a VIES-el. A VATAPI-s megoldást vizsgáljuk, de lehet, hogy ott külön csomagot kéne kialakítani a providernek, mert ha bekötöm az összes rendszert oda, akkor hamar elfogy a havi 150k lekérdezés :) 

Előnyös lenne mindenképp egy ilyen endpoint a navnál is, más alkalmazások fejlesztői is örülnének neki, nem beszélve arról, hogy nem menne be invalid adószám a NAV-hoz.

**Reply by vectorkovacspeter** [reply]
VIES -el nekem az volt még a gondom, hogy random nem volt hajlandó kiszolgálni bizonyos IP-ket, csak üzenetet küldött, hogy írjak az adminisztrátoroknak, akik aztán soha nem írtak vissza, vagy nem tudtak segíteni. Egy gazdátlan és instabil dolognak tűnt az m2m interfészük sajnos.

**Reply by robertsimon** [reply]
Nálunk is ez volt, aztán küldtünk a VIES supportnak egy naaaagy pecsétes, céges nyilatkozatot, hogy whitelisteljék a cég (statikus) IP címeit, mert onnan jönnek kérések, mert tax compliance-et csinálunk nem keveset az ügyfeleinknek. 
Azóta max csak olyan van, hogy lassítanak, de azt meg figyeli egy kódom, ami beiktat a következő batch küldésnél egy chill down 2 percet. 
Ettől függetlenül nagyon rossz a validációs rate, tegnap pl. 13k adószámot nyomtam el feléjük 1000-es batchekben, aminek kb 53-61%-volt sikeres validálás. A többit viszont több iterációba kellett újraküldeni feléjük, úgyhogy elment a nap vele.  Meg ugye egyenként kell küldeni (persze ott a Paralell.Task() és a sok kis validátor node). 
Szerintem jó lenne egy ilyen szolgáltatás a NAV-tól is, főleg ha támogatnák a batch validálást is. Azonban nyilván ez fejlesztés részükről is, főleg, ha nem VIES-en keresztül mennek.

**Comment by renced42** [comment]
Zárom a kérdést, most ilyn szolgáltatásunk nincsen, de cimkéztem új igénynek.
